geode-core/src/main/java/org/apache/geode/cache/query/internal/index/AbstractIndex.java

```diff
@@ -1826,9 +1828,14 @@ void addValuesToCollection(Collection result, CompiledValue iterOp, RuntimeItera
                     // Compare the value in index with value in RegionEntry.
                     ok = verifyEntryAndIndexValue(entry, o1, context);
                   }
-                  if (ok && runtimeItr != null) {
-                    runtimeItr.setCurrent(o1);
-                    ok = QueryUtils.applyCondition(iterOp, context);
+                  try {
+                    if (ok && runtimeItr != null) {
+                      runtimeItr.setCurrent(o1);
+                      observer.beforeIterationEvaluation(iterOp, o1);
+                      ok = QueryUtils.applyCondition(iterOp, context);
+                    }
+                  } finally {
+                    observer.afterIterationEvaluation(ok);
```

Comment:
I think you should try to extract the blocks that look like this to their own private methods. There is some repetition involved that you may be able to reduce down to one method. The code calling these methods will also become more readable.